controller/lib/pid/pid.cpp

```diff
@@ -65,11 +64,18 @@ bool PID::Compute() {
   unsigned long now = Hal.millis();
   unsigned long timeChange = (now - lastTime);
   if (timeChange >= SampleTime) {
+    // compute actual samples time-difference to take jitter into account in
+    // integral and derivative
+    double samplesTimeChangeSec = ((now - lastSampleTime) / 1000.0);
+    if (samplesTimeChangeSec <= 0) {
+      // something went wrong, assume sample time
+      samplesTimeChangeSec = (SampleTime / 1000.0);
+    }
     /*Compute all the working error variables*/
     double input = *myInput;
     double error = *mySetpoint - input;
-    double dInput = (input - lastInput);
-    outputSum += (ki * error);
+    double dInput = ((input - lastInput) / samplesTimeChangeSec);
```

Comment:
Is it at all possible that samplesTimeChangeSec is 0?

Reply:
I capture the case at line 70 of changed file:
    double samplesTimeChangeSec = ((now - lastSampleTime) / 1000.0);
    if (samplesTimeChangeSec <= 0) {
      // something went wrong, assume sample time
      samplesTimeChangeSec = (SampleTime / 1000.0);
    }

An alarm should be raised if this happens though...

Reply:
Changed the behaviour and actually don't run if no time has passed since last call instead of assuming sample time. This is safer and should not bring any problem.